### ISSUE TITLE
Logging metrics factory

### ISSUE THREAD

**Issue by glbrntt** [apple/swift-metrics#70]
When adding metrics to an application or framework it would be useful to have some indication of the metrics emitted during execution to serve as a basic sanity check ("when I do this does metric X get emitted?"). 

Providing a metrics factory which prints to stdout/stderr whenever a value is recorded would be helpful for this case. Is this something we think we should provide in `swift-metrics`?

**Comment by ktoso** [comment]
Sounds like a useful one -- I have a similar thing in my projects where it can capture the emitted metrics and assertions can be performed on them.

I do agree that it's useful, but I'm also on the side that swift-metrics (and swift-log) should contain no implementations, all backends should be external libs, even if small ones IMHO and these libs are best to remain as little impl and opinionated as possible (main reason being that this is a tremendously slippery slope, once we add a small thing, and another small thing, people get used to "the lib has things" and more and more stuff gets added).

I'm supportive, but would love to find a way we can put those not as the part of "the API project". If it's too difficult for such under the apple/ namespace perhaps we could use the swift-server org or some other idea?

**Comment by MrLotU** [comment]
Fully agree with @ktoso here.

Great idea but probably not best suited to be included in the `swift-metrics` project directly

**Comment by glbrntt** [comment]
> I do agree that it's useful, but I'm also on the side that swift-metrics (and swift-log) should contain no implementations, all backends should be external libs, even if small ones IMHO and these libs are best to remain as little impl and opinionated as possible (main reason being that this is a tremendously slippery slope, once we add a small thing, and another small thing, people get used to "the lib has things" and more and more stuff gets added).

I totally get this and agree it's a slippery slope.

My concern is that putting this in another package would introduce a level of friction that would render it pointless: I wouldn't want to release a framework which depended on it (even if it was only in tests) and I probably wouldn't manually add/remove it from my package manifest whenever I wanted to do a quick sanity check. In all likelihood I'd end up copying and pasting it whenever I needed it.

**Comment by glbrntt** [comment]
@ktoso I was thinking about this a little over the weekend: maybe a more general purpose metrics-testing package could be a home for this.

**Comment by ktoso** [comment]
Hey there, sorry was offline for the last week.

You mean as separate library but inside this repo, and call it some `EventCaptureTestLogHandler`? (the impl I use for logging I call "LogCapture" I guess, so thinking if the word capturing could make sense for the family of such handlers... the same for metrics/tracing/logging)

I still think it's a slippery slope to add more things here rather than make many smaller repos -- smaller repos put these "on level playing field" as any other person's implementation and show more intent that yes, there should be tons of implementations by various people, rather than in this repo as "the" one... but others maybe disagree. I'm kind of hoping it's not hard for us to spin new repos in swift-server org perhaps if we want to make such handlers available?

@weissi @tomerd wdyt?

**Comment by glbrntt** [comment]
I actually meant in a separate repo :)

**Comment by ktoso** [comment]
Oh awesome :-) Then I'm super supportive of it.
Others may want to chime in, but sounds great to me.

**Comment by MrLotU** [comment]
Having a `swift-metrics-tools` kind of repository in the `swift-server` org sounds like a great idea to me. 

👍

**Comment by tomerd** [comment]
sorry for the slow response here, for whatever reasons my email filtered our this thread. 

I agree with everything that was said here. I also find myself needing this often in debugging context so would be nice to have a utility to bridge metrics into logs in that context. I also agree that a separate package/repo is the way to go. @glbrntt is this something you want to kick off?

**Comment by glbrntt** [comment]
Yeah, I'm happy to do that; I'll pitch something to the forums to help shape this.

**Comment by glbrntt** [comment]
fyi: https://forums.swift.org/t/metrics-extras/36616
